library/core/src/iter/adapters/enumerate.rs

```diff
@@ -15,10 +16,131 @@ use crate::ops::{Add, AddAssign, Try};
-        Enumerate { iter, count: 0 }
+        EnumerateImpl::new(iter)
+    }
+}
+
+// Enumerate specialization trait
+#[doc(hidden)]
+trait EnumerateImpl<I> {
+    type Item;
+    fn new(iter: I) -> Self;
+    fn next(&mut self) -> Option<Self::Item>;
+    unsafe fn __iterator_get_unchecked(&mut self, idx: usize) -> Self::Item
+    where
+        Self: TrustedRandomAccess;
+    fn next_back(&mut self) -> Option<Self::Item>
+    where
+        I: ExactSizeIterator + DoubleEndedIterator;
+}
+
+impl<I> EnumerateImpl<I> for Enumerate<I>
+where
+    I: Iterator,
+{
+    type Item = (usize, I::Item);
+
+    default fn new(iter: I) -> Self {
+        Enumerate {
+            iter,
+            count: 0,
+            len: 0, // unused
+        }
+    }
+
+    #[inline]
+    default fn next(&mut self) -> Option<Self::Item> {
+        let a = self.iter.next()?;
+        let i = self.count;
+        // Possible undefined overflow.
+        AddAssign::add_assign(&mut self.count, 1);
+        Some((i, a))
+    }
+
+    #[inline]
+    default unsafe fn __iterator_get_unchecked(&mut self, idx: usize) -> Self::Item
+    where
+        Self: TrustedRandomAccess,
+    {
+        // SAFETY: the caller must uphold the contract for
+        // `Iterator::__iterator_get_unchecked`.
+        let value = unsafe { try_get_unchecked(&mut self.iter, idx) };
+        (Add::add(self.count, idx), value)
+    }
+
+    #[inline]
+    default fn next_back(&mut self) -> Option<Self::Item>
+    where
+        I: ExactSizeIterator + DoubleEndedIterator,
+    {
+        let a = self.iter.next_back()?;
+        let len = self.iter.len();
+        // Can safely add, `ExactSizeIterator` promises that the number of
+        // elements fits into a `usize`.
+        Some((self.count + len, a))
+    }
+}
+
+// This is the same code as above but using `intrinsics::assume()` to hint at the compiler
+// that the returned index is smaller than the length of the underlying iterator.
+impl<I> EnumerateImpl<I> for Enumerate<I>
+where
+    I: TrustedLen + Iterator,
+{
+    fn new(iter: I) -> Self {
+        let len = iter.size_hint().0;
+
+        Enumerate { iter, count: 0, len }
+    }
+
+    #[inline]
+    fn next(&mut self) -> Option<Self::Item> {
+        let a = self.iter.next()?;
+        // SAFETY: There must be fewer than `self.len` items because of `TrustedLen`'s API contract
+        unsafe {
+            intrinsics::assume(self.count < self.len);
+        }
+        let i = self.count;
+        // Possible undefined overflow.
+        AddAssign::add_assign(&mut self.count, 1);
+        Some((i, a))
+    }
+
+    #[inline]
+    unsafe fn __iterator_get_unchecked(&mut self, idx: usize) -> Self::Item
+    where
+        Self: TrustedRandomAccess,
+    {
+        // SAFETY: the caller must uphold the contract for
+        // `Iterator::__iterator_get_unchecked`.
+        let value = unsafe { try_get_unchecked(&mut self.iter, idx) };
+        let idx = Add::add(self.count, idx);
```

Comment:
Why does this use `Add::add` instead of `+`? If this is necessary, a comment explaining why would be useful.

Reply:
That's how it was [originally](https://github.com/rust-lang/rust/blob/923e3d2400e9a56fed70d377db5e7728ed3e5192/library/core/src/iter/adapters/enumerate.rs#L120) and is everywhere else in this file. I don't know why and was wondering the same, but kept it because there presumably was a reason

Reply:
The `Add::add` delays decision about overflow check mode to the crate that does the code generation, even if overflow checks are disabled in the current crate, which they usually are for the standard library.

Reply:
According to https://github.com/rust-lang/rust/issues/81721, we should be using `#[rustc_inherit_overflow_checks]` with a regular `+` operator here instead of the `Add:add` trick.